File: src/main/java/com/graphbrain/db/Graph.java
package com.graphbrain.db;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class Graph {

	private Backend back;
	
	public Graph() {
		back = new LevelDbBackend();
	}
	
	public Vertex get(String id) {
		return back.get(id, VertexType.type(id));
	}
	
	public Vertex put(Vertex vertex) {
		back.put(vertex);
		
		if (vertex.type() == VertexType.Edge) {
			onPutEdge((Edge)vertex);
		}
		
		return vertex;
	}
	
	public Vertex update(Vertex vertex) {
		return back.update(vertex);
	}
	
	public void remove(Vertex vertex) {
		back.remove(vertex);
		
		if (vertex.type() == VertexType.Edge) {
			onRemoveEdge((Edge)vertex);
		}
	}
	
	public boolean exists(String id) {
		return get(id) != null;
	}
	
	public UserNode getUserNode(String id) {
		return (UserNode)back.get(id, VertexType.User);
	}
	
	private String idFromEmail(String email) {
		String userName = back.usernameByEmail(email);
		if (userName == null) {
			return null;
		}
		else {
			return ID.idFromUsername(userName);
		}
	}
	
	public boolean usernameExists(String username) {
		return exists(ID.idFromUsername(username));
	}
	
	public boolean emailExists(String email) {
		String userName = back.usernameByEmail(email);
	    return userName != null; 
	}
	
	public UserNode findUser(String login) {
		if (exists(ID.idFromUsername(login))) {
		  	  return getUserNode(ID.idFromUsername(login));
		}
		else {
			String uid = idFromEmail(login);
		    if (uid == null) {
		    	return null;
		    }
		    else if (exists(uid)) {
		  	    return getUserNode(idFromEmail(login));
		    }
		    else {
		        return null;
		    }
		}
	}
	
	public UserNode getUserNodeByUsername(String username) {
		if (exists(ID.idFromUsername(username))) {
			return getUserNode(ID.idFromUsername(username));
		}
		else {
			return null;
		}
	}
	
	public UserNode createUser(String username, String name, String email, String password, String role) {
		UserNode userNode = UserNode.create(username, name, email, password, role);
		back.put(userNode);
		if (!email.equals("")) {
		    back.associateEmailToUsername(email, username);
		}

		return userNode;
	}
	
	public UserNode attemptLogin(String login, String password) {
		UserNode userNode = findUser(login);

		// user does not exist
		if (userNode == null) {
			return null;
		}
		    
		// password is incorrect
		if (!userNode.checkPassword(password)) {
			return null;
		}

		// ok, create new session
		userNode.newSession();
		back.update(userNode);
		return userNode;
	}
	
	public UserNode forceLogin(String login) {
		UserNode userNode = findUser(login);

		// user does not exist
		if (userNode == null) {
			return null;
		}

		// ok, create new session
		userNode.newSession();
		back.update(userNode);
		return userNode;
	}
	
	public List<Vertex> allUsers() {
		return back.listByType(VertexType.User);
	}
	
	public Set<Edge> edges(Vertex center) {
		return back.edges(center);
	}
	
	public Set<Edge> edges(String centerId) {
		Vertex center = get(centerId);
		return edges(center);
	}
	
	public Set<String> nodesFromEdgeSet(Set<Edge> edgeSet) {
		Set<String> nset = new HashSet<String>();

		for (Edge e : edgeSet) {
			for (String pid : e.getIds()) {
				nset.add(pid);
		    }
		}
		
		return nset;
	}

	public Set<String> neighbors(String centerId) {
		//ldebug("neighbors " + nodeId)

		Set<Edge> nedges = edges(centerId);
		Set<String> nodes = nodesFromEdgeSet(nedges);
		nodes.add(centerId);
		return nodes;
	}
	
	protected void incDegree(Vertex vertex) {
		vertex.incDegree();
		update(vertex);
	}
	
	protected void incDegree(String id) {
		Vertex vertex = get(id);
		incDegree(vertex);
	}
	
	protected void decDegree(Vertex vertex) {
		vertex.decDegree();
		update(vertex);
	}
	
	protected void decDegree(String id) {
		Vertex vertex = get(id);
		decDegree(vertex);
	}
	
	protected void onPutEdge(Edge edge) {
		for (String id : edge.getIds()) {
			incDegree(id);
		}
	}
	
	protected void onRemoveEdge(Edge edge) {
		for (String id : edge.getIds()) {
			decDegree(id);
		}
	}

/*

  def addrel(edge: Edge): Edge = {
    ldebug("addrel " + edge)

    if (!relExists(edge)) {
      incInstances(edge.edgeType)
      for (i <- 0 until edge.participantIds.size) {
        val p = edge.participantIds(i)
        addEdgeEntry(p, edge)
        incVertexEdgeType(p, edge.edgeType, i)
        incDegree(p)
      }
    }

    edge
  }


  def addrel(edgeType: String, participants: List[String]): Edge = addrel(Edge(edgeType, participants))


  def delrel(edge: Edge): Unit = {
    ldebug("delrel " + edge)

    if (relExists(edge)) {
      decInstances(edge.edgeType)
      for (i <- 0 until edge.participantIds.size) {
        val p = edge.participantIds(i)
        delEdgeEntry(p, edge)
        decVertexEdgeType(p, edge.edgeType, i)
        decDegree(p)
      }
    }
  }


  def delrel(edgeType: String, participants: List[String]): Unit = delrel(Edge(edgeType, participants))


  def createAndConnectVertices(edgeType: String, participants: Array[Vertex]) = {
    ldebug("createAndConnectVertices edgeType: " + edgeType + "; participants: " + participants)
    for (v <- participants) {
      if (!exists(v.id)) {
        put(v)
      }
    }

    val ids = for (v <- participants) yield v.id
    addrel(edgeType.replace(" ", "_"), ids.toList)
  }
  */
}

File: core/src/main/scala/com/graphbrain/db/Graph.scala
package com.graphbrain.db

import java.util.Date
import com.typesafe.scalalogging.slf4j.Logging

class Graph(name: String="dbnode") extends Logging {
  val back = new LevelDbBackend(name)

  def get(id: String): Vertex = back.get(id, VertexType.getType(id))

  def put(vertex: Vertex): Vertex = {
    val v = if (vertex.ts < 0)
      vertex.setTs(new Date().getTime)
    else
      vertex

    back.put(v)

    v match  {
      case e: Edge => onPutEdge(e)
      case _ =>
    }

    v
  }

  def update(vertex: Vertex): Vertex = put(vertex)

  def exists(id: String): Boolean = get(id) != null

  def exists(v: Vertex): Boolean = exists(v.id)

  def remove(vertex: Vertex) = {
    back.remove(vertex)

    vertex match {
      case e: Edge => onRemoveEdge(e)
      case _ =>
    }
  }

  def connectVertices(participants: Array[String]) = {
    logger.debug(s"connectVertices participants: ${participants.mkString(" ")}")
    put(Edge.fromParticipants(participants))
  }

  def edges(pattern: Edge) = back.edges(pattern)

  def edges(center: Vertex) = back.edges(center)

  def edges(centerId: String): Set[Edge] = edges(get(centerId))

  def edges(pattern: Array[String]) = back.edges(Edge.fromParticipants(pattern))

  def nodesFromEdgeSet(edgeSet: Set[Edge]) = {
    var nset = Set[String]()

    for (e <- edgeSet)
      for (pid <- e.ids)
        nset = nset + pid

    nset
  }

  def neighbors(centerId: String) = {
    logger.debug(s"neighbors: $centerId")

    val nedges = edges(centerId)
    val nodes = nodesFromEdgeSet(nedges)
    nodes + centerId
  }

  protected def incDegree(vertex: Vertex): Vertex = update(vertex.setDegree(vertex.degree + 1))

  protected def incDegree(id: String): Vertex = incDegree(get(id))

  protected def onPutEdge(edge: Edge) =
    for (id <- edge.ids) {
      if (!exists(id)) {
        put(Vertex.createFromId(id))
      }
      incDegree(id)
    }

  protected def decDegree(vertex: Vertex): Vertex = update(vertex.setDegree(vertex.degree - 1))

  protected def decDegree(id: String): Vertex = {
    val v = get(id)
    if (v != null) decDegree(v) else null
  }

  protected def onRemoveEdge(edge: Edge) =
    for (id <- edge.ids)
      decDegree(id)

  def description(vertex: Vertex): String = {
    val asIn = edges(Array[String]("rtype/1/as_in", vertex.id, "*"))

    vertex.toString + (if (asIn.size == 0) {
      ""
    }
    else {
      " (" + asIn.map(e => get(e.participantIds(1)).toString).mkString(", ") + ")"
    })
  }

  def description(id: String): String = description(get(id))

  def getProgNode(id: String): ProgNode = get(id) match {
    case p: ProgNode => p
    case _ => null
  }

  def getTextNode(id: String): TextNode = get(id) match {
    case t: TextNode => t
    case _ => null
  }
}